### DISCUSSION TITLE
Tesla API went down , CAPTCHA installed.

### DISCUSSION THREAD

**Discussion by fkhera** [timdorr/tesla-api#393] [Q&A]
Has anyone had issues with the Tesla API, our raspberry PI which just accesses the token , oauth API seems to have gone down.

Code used for REST API:
https://auth.tesla.com/oauth2/v3/authorize?code_challenge=Q9MaFZwd3Y0GQx1E42nGDn4NtnAhpgqh6l-BwdYe6Tg&client_id=ownerapi&state=tesla_exporter&redirect_uri=https%3A%2F%2Fauth.tesla.com%2Fvoid%2Fcallback&response_type=code&code_challenge_method=S256&scope=openid+email+offline_access

IT appears Tesla is using some type "code catchya" , any ideas how to get past that?
Following the link above shows a captchya.

Coming to non MFA flow:
Traceback (most recent call last):
  File "/home/pi/powerwallTimeOfUse.py", line 242, in authenticate
    code_url = resp.headers["location"]
  File "/usr/lib/python2.7/dist-packages/requests/structures.py", line 52, in __getitem__
    return self._store[key.lower()][1]
KeyError: 'location'
Traceback (most recent call last):
  File "/home/pi/powerwallTimeOfUse.py", line 67, in main
    tpw.productlist()

**Comment by fkhera** [comment]
A bit of a long shot but could use human solving service:
https://solvecaptcha.com/api-description

https://auth.tesla.com/oauth2/v1/authorize?client_id=teslaweb&response_type=code&scope=openid%20email%20profile&redirect_uri=https%3A//www.tesla.com/openid-connect/generic&state=_WDi4NXQGYV7yGOi9df-2qGznv9xzaUz2RUKEIo0dpU&locale=en-US

It appears the new API piece is:

_csrf: rPIvYAwz-131lLqcOXxNPeZmUP-BZzyiGB7k
_phase: authenticate
_process: 1
transaction_id: aQk9oHMs
cancel: 
identity: email
credential: pass
captcha: Pwab

**Reply by hemang249** [reply]
If your service is an app, you could use a embedded web view to show the /authorize tesla page in the app and capture the oauth token once it tries to redirect to /void/callback

**Reply by themonomers** [reply]
You can write the CAPTCHA image to your Google Drive, view it there, prompt the user for the text in Python, and then continue on with the flow.  

[Github](https://github.com/themonomers/tesla/blob/master/python/TeslaToken.py)

**Comment by carlosonunez** [comment]
I'm also encountering this problem.

**Comment by carlosonunez** [comment]
Unfortunately that won’t work given that I autogenerated these tokens every 45 days through GitHub Actions. :(

**Comment by fkhera** [comment]
If anyone is still running into this problem and using python they can see
how we implemented captcha check and resolve it.
https://github.com/fkhera/powerwallCloud/commit/ba4012a4d6b6cfd0f5fd8d8299577fca005bb42f

It uses 2catpcha, to farm off the image, which is saved as .svg , converted
.png.
Call 2captcha to have them solve.
Wait on their resolution.
Use the resolution : captcha code, and embed it into Tesla request.

And continue the rest of the flow as prescribed.  From what I can see the
code is working as expected.

**Comment by magdawg** [comment]
@fkhera I can see in your code you obtain the `code` from the /authorize endpoint response by querying for https://github.com/fkhera/powerwallCloud/blob/e67980a4febdda07b79db5b5539e47bbfecdbd2a/powerwallBackup.py#L258

In my response, the Location header does not have a key `code`, only `code-challenge`. Any chance you encountered this while working on your solution?

I continue with my `code-challenge` value from the Location header to the /token endpoint and get `invalid_grant` error.

**Comment by fkhera** [comment]
Are you using captcha ??

**Reply by magdawg** [reply]
Yes. I provide `"captcha": "<value_from_captcha_svg>"` in the body of the request to the `/authorize` endpoint. And it returns a response with bunch of headers, including the Location header that has the redirect URI but no `code` in it. The body of the response is something like:
```
Found. Redirecting to https://auth.tesla.com/oauth2/v3/authorize?audience=&client_id=ownerapi&code_challenge=-nZs9VDs3A_BMWn2QyjKLnU1B0SpAev57n-uozj6O6w&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fauth.tesla.com%2Fvoid%2Fcallback&response_type=code&scope=openid+email+offline_access&state=8zeITVpNtluR97mO1f8HSA
```
